snuba/utils/streams/kafka.py

```diff
@@ -120,22 +121,85 @@ def close(self) -> None:
     queued_min_messages: int = DEFAULT_QUEUED_MIN_MESSAGES,
-) -> KafkaConsumer:
-    return KafkaConsumer(
-        {
-            "enable.auto.commit": False,
-            "bootstrap.servers": ",".join(bootstrap_servers),
-            "group.id": group_id,
-            "default.topic.config": {"auto.offset.reset": auto_offset_reset},
-            # overridden to reduce memory usage when there's a large backlog
-            "queued.max.messages.kbytes": queued_max_messages_kbytes,
-            "queued.min.messages": queued_min_messages,
-            "enable.partition.eof": False,
-        }
+) -> Mapping[str, Any]:
+    return {
+        "enable.auto.commit": False,
+        "bootstrap.servers": ",".join(bootstrap_servers),
+        "group.id": group_id,
+        "default.topic.config": {"auto.offset.reset": auto_offset_reset},
+        # overridden to reduce memory usage when there's a large backlog
+        "queued.max.messages.kbytes": queued_max_messages_kbytes,
+        "queued.min.messages": queued_min_messages,
+        "enable.partition.eof": False,
+    }
+
+
+class KafkaConsumerWithCommitLog(KafkaConsumer):
+
+    # Set of logical (not literal) offsets to not publish to the commit log.
+    # https://github.com/confluentinc/confluent-kafka-python/blob/443177e1c83d9b66ce30f5eb8775e062453a738b/tests/test_enums.py#L22-L25
+    LOGICAL_OFFSETS = frozenset(
+        [OFFSET_BEGINNING, OFFSET_END, OFFSET_STORED, OFFSET_INVALID]
     )
+
+    def __init__(
+        self,
+        configuration: Mapping[str, Any],
+        producer: ConfluentProducer,
+        commit_log_topic: str,
+    ) -> None:
+        super().__init__(configuration)
+        self.__producer = producer
+        self.__commit_log_topic = commit_log_topic
+        self.__group_id = configuration["group.id"]
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[KafkaMessage]:
+        self.__producer.poll(0.0)
+        return super().poll(timeout)
+
+    def __commit_message_delivery_callback(
+        self, error: Optional[KafkaError], message: ConfluentMessage
+    ) -> None:
+        if error is not None:
+            raise Exception(error.str())
+
+    def commit(self) -> Mapping[TopicPartition, int]:
+        offsets = super().commit()
+
+        if self.__commit_log_topic:
+            for stream, offset in offsets.items():
+                # XXX: This is an abstraction leak from the Kafka consumer.
+                if offset in self.LOGICAL_OFFSETS:
+                    logger.debug(
+                        "Skipped publishing logical offset (%r) to commit log for %s",
+                        offset,
+                        stream,
+                    )
+                    continue
+                elif offset < 0:
+                    logger.warning(
+                        "Found unexpected negative offset (%r) after commit for %s",
+                        offset,
+                        stream,
+                    )
```

Comment:
I think this should probably be moved to the parent class definition. These logical offsets are a implementation detail of the Confluent client and probably shouldn't be exposed as part of the public API via the `commit` return value. We should probably just document that the set of streams returned by this method is a potentially subset of the streams that are in the subscription set.

Reply:
I did move this to the superclass with 72c5535.

I am pretty sure this only occurs if the consumer is subscribed to multiple partitions (or the consumer implementation may have changed since we originally wrote this.) Testing this with a single partition before any work is committed yields this error:

```
Traceback (most recent call last):
  File "/Users/ted/workspaces/snuba/tests/utils/streams/test_kafka.py", line 63, in test_consumer
    assert consumer.commit() == {}
  File "/Users/ted/workspaces/snuba/snuba/utils/streams/kafka.py", line 107, in commit
    result = self.__consumer.commit(asynchronous=False)
cimpl.KafkaException: KafkaError{code=_NO_OFFSET,val=-168,str="Commit failed: Local: No offset stored"}
```

Reply:
I'm not sure we want to throw in that situation anyway (calling commit on a slow moving topic seems like it's not that big of a deal) so we might want to catch and discard that exception anyway?